apps/webapp/src/components/swap/unclaimed-swaps.tsx

```diff
@@ -0,0 +1,75 @@
+import { Button, Card } from '@penumbra-zone/ui';
+import { useLoaderData, useRevalidator } from 'react-router-dom';
+import { SwapLoaderResponse } from './swap-loader.tsx';
+import { AssetIcon } from '../shared/asset-icon.tsx';
+import { issueSwapClaim } from '../../state/swap.ts';
+import { SwapRecord } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/view/v1/view_pb';
+import { getSwapRecordCommitment } from '@penumbra-zone/types';
+import { toast, useToast } from '@penumbra-zone/ui/components/ui/use-toast.ts';
+import { errorTxToast, loadingTxToast, successTxToast } from '../shared/toast-content.tsx';
+import { Dispatch, SetStateAction, useState } from 'react';
+
+const claimSwap = async (
+  swap: SwapRecord,
+  toastFn: typeof toast,
+  setInProgress: Dispatch<SetStateAction<boolean>>,
+  reloadData: () => void, // Used to refresh page state after action
+) => {
+  const { dismiss } = toastFn(loadingTxToast);
+  setInProgress(true);
+
+  try {
+    const commitment = getSwapRecordCommitment(swap);
+    const swapClaimTxHash = await issueSwapClaim(commitment);
+
+    dismiss();
+    toastFn(successTxToast(swapClaimTxHash));
+  } catch (e) {
+    toastFn(errorTxToast(e));
+    throw e;
+  } finally {
+    dismiss();
+    setInProgress(false);
+    reloadData();
+  }
+};
+
+export const UnclaimedSwaps = () => {
+  const { toast } = useToast();
+  const { revalidate } = useRevalidator();
+  const [inProgress, setInProgress] = useState(false);
+  const { unclaimedSwaps } = useLoaderData() as SwapLoaderResponse;
+  if (!unclaimedSwaps.length) return <></>;
+
+  return (
+    <Card className=''>
+      <p className='bg-text-linear bg-clip-text font-headline text-xl font-semibold leading-[30px] text-transparent md:text-2xl md:font-bold md:leading-9'>
+        Unclaimed Swaps
+      </p>
+      <p className='text-gray-400'>
+        Swaps on Penumbra are a two step process. The first transaction issues the request and the
+        second claims the result of the swap action. For some reason, these second transactions were
+        not completed. Claim them!
+      </p>
+      {unclaimedSwaps.map(({ swap, asset1, asset2 }, i) => (
+        <div key={i} className='mt-4 flex justify-between'>
+          <div className='flex flex-col gap-2'>
+            <div>Block Height: {Number(swap.outputData?.height)}</div>
+            <div className='flex items-center gap-2'>
+              <AssetIcon metadata={asset1} />
+              <span>↔</span>
+              <AssetIcon metadata={asset2} />
+            </div>
+          </div>
+          <Button
+            className='w-20'
+            onClick={() => void claimSwap(swap, toast, setInProgress, revalidate)}
+            disabled={inProgress}
+          >
+            Claim
```

Comment:
What do you think of:
```suggestion
            {inProgress ? 'Claiming' : 'Claim'}
```

I tried this locally, where I had two unclaimed swaps. Both buttons got greyed out and when I pressed "Claim," so I was wondering if it was claiming both of them simultaneously — although it became clear once it finished that only one of them was being claimed.

For that matter, do _all_ Claim buttons need to be disabled while any _one_ swap is being claimed? Can't we just disable the button next to the swap that's currently in progress, so that the user can click a bunch and have a bunch of them go in parallel? (We could keep track of which ones are in progress using e.g., the base 64-encoded swap commitment bytes as their IDs?)

Reply:
Yeah, this is probably a better user experience. Will create a new slice for this logic.